### PR TITLE
Fix wall negative height

### LINKED ISSUE #4274: Wall geometries with all negative heights do not render
I need to represent wall geometries close to the ground surface with terrain turned on.

due to the geoid offset at particular locations the elevation for the top of the wall will be a negative value for the wall geometries since cesium is using ellipsoid elevations rather than orthometric.

a working example with two walls, the top of the red wall just above 0

https://sandcastle.cesium.com/#c=3ZNvT9swEMa/yilvSKXgJIOxrQvVSkGaNKQhxsYLwgs3Plpr/lOdnXRl4rvjJIUVOmlM2qs5kpU7/567ix+l4QSNxCUSHILBJUzQyVqzb10uLqOqiyfWeC4NUhkl8LM0AB6JQuaMbCMF0vBBWBFyj5eWlLjokXiQlOZu8L40/dOEllNV4yVXKjTtuzM0XnqJjnEh4q4DgOEaYQg7R4GGZYvvJP1JFwxhzbVrYV3QW+PgcZQJJx/euNljN2T1Mc4I0Y3DUKv4ajd/tceygzcJ7L9jeb6f/Cr1l2td6e260sH1YKOW5j+krvVHlLO5b2e7yrME8gyuNyFpnkGQsTxptydYuFiSXG18oVWW2NHp1xO2lH4+Vos5jzP2egAbMlt7Fazr2G3pePKpR++2TCIUL/foHEXnyn/gUMay9u7/5NFu62TYXmLR+cnxv3Ro7cettfrCxs/caZEoiQrnVwpHpfkg9cKSh5pUzFjqUS9UGNOl07r6jp5VrpMU6YOgELIBKQ5/8/NDpbhz4eSmVuqLvMUyGhVp4J/IlOVCmtnnBknxVYvM89Fpn2SMFWkIt1XeWjXl9FjxHg

for the red wall in the provided example change  the maximumHeights  line 
from: 
maximumHeights : [0.01, 0.01 ],

to
maximumHeights : [-0.01, -0.01 ],

the red wall no longer renders

### PR DESCRIPTION
**Pull request by baothientran**
Fixes #4274 

Currently there is a check to prevent maxHeights from being negative in the function `removeDuplicates()` in WallGeometryLibrary.js file. I remove it, and walls with negative height are rendered correctly. But with the old method when the first member of maximumHeights array is equal to 0.0, normal vector can't be calculated. It's because vertices of the geometry are clamped to the earth's surface first before being used to calculate normal. So when the vertices themselves are on the surface, they and the clamped values have the same coordinates, which leads to undefined normal vectors. for example, there is an exception thrown when running this [Sandcastle example](http://localhost:8080/Apps/Sandcastle/index.html#c=jZLPT8IwFID/lZedRgIdEzyIg2jAxIOJCRo9MA5le0Bjf5C2G4Lhf7fdIC7IgR26vtfve2v7VlINJcMtahiCxC2M0bBCkI8qF6ZBVsVjJS1lEnUatO5TmcrSeRrzT8q5E+sKBKVllqEhNM/Dn1QCSCpwAGkwxRy2nqUW1shWa5sGbQ/45AAqFmCjjPOVNIPTNsZUWzejskeWWokJrjSiedSa7p6rMiac1S5AJ45vSbd9Cvv9RnDT9U8j0bnrXs/2rmXnreNE0G8mCnHc4wBmnqlo/46P1vxEM3mBjmu4yVGLmlH+dz2KK02mT5MKOLjxcGxP3ZC9UuJdhWft8UjQDhJjdxxHp6M8MLFR2kKheUhIZFFsuPueiRZF9oWWZKYSPZpETTXJWQksH174VyDj1Bi3siw4f2N7TINREjn+n8oVzZlcvZaoOd15bB2PXuokISSJXHjZtErxBdVnlX8B). I'm not sure if we should let maximumHeight == minimumHeight. But since it works with value other than 0.0 in the master branch, I think I should fix it as well.

For the undefined normal vector problem above, instead of clamping the top vertex to the earth surface, I use the bottom vertex to calculate normal. If the top and bottom positions have the same coordinates at the beginning (aka maximumHeight[0] == minimumHeights[0]), I just assign Y_Axis to normal, X_Axis to tangent, and Z_Axis to bitangent. If the top and bottom positions in the middle of the wall are overlapped, I just assigned the last valid normal to it. Same with tangent and bitangent. So that should prevent all of those vectors from having zero length, which leads to the error above. I had to cleanup the WallGeometry.createGeometry() method more than I expected to accommodate the changes better. Please let me know if you have a better way.

There is also a bug where texture coordinate seems to be larger than 1.0 for [this sandcastle's example](http://localhost:8080/Apps/Sandcastle/index.html?src=development%2FGeometry%20and%20Appearances.html). This is the wall with checkerboard texture rendered in the master branch:

Screen Shot 2020-07-07 at 10 43 32 PM

This is the fix in the PR:

Screen Shot 2020-07-07 at 10 45 03 PM

@hpinkos @lilleyse Can you please take a look at it? Please let me know if there are anything I need to test

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @baothientran!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/fix-wall-negative-height/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

File: Specs/Core/WallGeometrySpec.js [hunk 1]
```diff
@@ -82,6 +82,53 @@ describe("Core/WallGeometry", function () {
...
+          -90.0,
+          44.0,
+          200000.0,
+          -30.0,
+          44.0,
+          200000.0,
+        ]),
+        maximumHeights: [40000.0, 60000.0, 60000.0, 50000.0],
+        minimumHeights: [60000.0, 60000.0, 60000.0, 50000.0],
+      })
```
Comment (baothientran) [inline Specs/Core/WallGeometrySpec.js:126]: I'm not sure if maximumHeights should always be larger than minimumHeights. This seems to work in the master, so I assume I should test as well

**Comment by hpinkos** [comment]
Thanks @baothientran!  I'll try to give this a review early next week

**Comment by lilleyse** [comment]
@hpinkos will you be able to look at this PR soon?

**Comment by hpinkos** [comment]
@lilleyse, yes, this and a few other CesiumJS PRs are on my list to review

**Comment by hpinkos** [comment]
@baothientran thanks for the PR, but the amount of re-factoring you did along the way made this impossible to review.  It's important that each PR only contain code changes related to a single issue instead of adding in changes to catch other things you find along the way.

I've opened a PR here to fix the negative height issue: https://github.com/CesiumGS/cesium/pull/9041
I've also opened this issue related to the incorrect texture coordinate: https://github.com/CesiumGS/cesium/issues/9042

It would be much appreciated if you could open a new PR for fixing https://github.com/CesiumGS/cesium/issues/9042.  Thanks!

**Pull request closed without merging**